File: perl/tests/test-radionuclides.pl
use strict;
use Test::More;
use xraylib;
use xrltest;

my $list = xraylib::GetRadioNuclideDataList();
ok(scalar @{$list} eq 10, "TestRadionuclides:test_good::list");
for my $i (0..9) {
	my $rnd = xraylib::GetRadioNuclideDataByIndex($i);
	ok($rnd->{name} eq $list->[$i], "TestRadionuclides:test_good::index");
	$rnd = xraylib::GetRadioNuclideDataByName($list->[$i]);
	ok($rnd->{name} eq $list->[$i], "TestRadionuclides:test_good::name");
}

my $rnd = xraylib::GetRadioNuclideDataByIndex(3);
ok($rnd->{'A'} eq 125, "TestRadionuclides::test_good_nr3::A");
ok($rnd->{'GammaEnergies'} ~~ [35.4919], "TestRadionuclides::test_good_nr3::GammaEnergies");
ok($rnd->{'GammaIntensities'} ~~ [0.0668], "TestRadionuclides::test_good_nr3::GammaIntensities");
ok($rnd->{'N'} eq 72, "TestRadionuclides::test_good_nr3::N");
ok($rnd->{XrayIntensities} ~~ [0.0023, 0.00112, 0.0063, 0.056, 0.035, 0.0042, 0.007, 0.00043, 0.0101, 0.0045, 0.00103, 0.0016, 3.24e-05, 0.406, 0.757, 0.0683, 0.132, 0.00121, 0.0381, 0.0058], "TestRadionuclides::test_good_nr3::XrayIntensities");
ok($rnd->{XrayLines} ~~ [-86, -60, -89, -90, -63, -33, -34, -91, -95, -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13], "TestRadionuclides::test_good_nr3::XrayLines");
ok($rnd->{'Z'} eq 53, "TestRadionuclides::test_good_nr3::Z");
ok($rnd->{'Z_xray'} eq 52, "TestRadionuclides::test_good_nr3::Z_xray");
ok($rnd->{'nGammas'} eq 1, "TestRadionuclides::test_good_nr3::nGammas");
ok($rnd->{'nXrays'} eq 20, "TestRadionuclides::test_good_nr3::nXrays");
ok($rnd->{'name'} eq "125I", "TestRadionuclides::test_good_nr3::name");

$rnd = xraylib::GetRadioNuclideDataByName("125I");
ok($rnd->{'A'} eq 125, "TestRadionuclides::test_good_125I::A");
ok($rnd->{'GammaEnergies'} ~~ [35.4919], "TestRadionuclides::test_good_125I::GammaEnergies");
ok($rnd->{'GammaIntensities'} ~~ [0.0668], "TestRadionuclides::test_good_125I::GammaIntensities");
ok($rnd->{'N'} eq 72, "TestRadionuclides::test_good_125I::N");
ok($rnd->{XrayIntensities} ~~ [0.0023, 0.00112, 0.0063, 0.056, 0.035, 0.0042, 0.007, 0.00043, 0.0101, 0.0045, 0.00103, 0.0016, 3.24e-05, 0.406, 0.757, 0.0683, 0.132, 0.00121, 0.0381, 0.0058], "TestRadionuclides::test_good_125I::XrayIntensities");
ok($rnd->{XrayLines} ~~ [-86, -60, -89, -90, -63, -33, -34, -91, -95, -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13], "TestRadionuclides::test_good_125I::XrayLines");
ok($rnd->{'Z'} eq 53, "TestRadionuclides::test_good_125I::Z");
ok($rnd->{'Z_xray'} eq 52, "TestRadionuclides::test_good_125I::Z_xray");
ok($rnd->{'nGammas'} eq 1, "TestRadionuclides::test_good_125I::nGammas");
ok($rnd->{'nXrays'} eq 20, "TestRadionuclides::test_good_125I::nXrays");
ok($rnd->{'name'} eq "125I", "TestRadionuclides::test_good_125I::name");


eval {
	xraylib::GetRadioNuclideDataByName('Uu');
};
like($@, qr/^ValueError/ , "TestRadionuclides::test_bad::name1");

eval {
	xraylib::GetRadioNuclideDataByName(undef);
};
like($@, qr/^ValueError/ , "TestRadionuclides::test_bad::name2");
eval {
	xraylib::GetRadioNuclideDataByIndex('Uu');
};
like($@, qr/^TypeError/ , "TestRadionuclides::test_bad::index1");
xraylib::GetRadioNuclideDataByIndex(undef); # undef gets translated to zero apparently!
eval {
	xraylib::GetRadioNuclideDataByIndex(-1);
};
like($@, qr/^ValueError/ , "TestRadionuclides::test_bad::index3");
eval {
	xraylib::GetRadioNuclideDataByIndex(180);
};
like($@, qr/^ValueError/ , "TestRadionuclides::test_bad::index4");


done_testing();

File: java/tests/TestRadioNuclides.java
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.stream.IntStream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.NullAndEmptySource;
 
import com.github.tschoonj.xraylib.Xraylib;
import com.github.tschoonj.xraylib.radioNuclideData;

public class TestRadioNuclides {

	@Test
	public void test_list_consistency() {
		String[] list = Xraylib.GetRadioNuclideDataList();
		assertEquals(list.length, 10);

		for (int i = 0 ; i < list.length ; i++) {
			radioNuclideData rnd = Xraylib.GetRadioNuclideDataByIndex(i);
			assertEquals(rnd.name, list[i]);
			rnd = Xraylib.GetRadioNuclideDataByName(list[i]);
			assertEquals(rnd.name, list[i]);
		}
	}

	@Test
	public void test_GetRadioNuclideDataByIndex_3() {
		radioNuclideData rnd = Xraylib.GetRadioNuclideDataByIndex(3);
		assertEquals(rnd.A, 125);
		assertArrayEquals(rnd.GammaEnergies, new double[]{35.4919});
		assertArrayEquals(rnd.GammaIntensities, new double[]{0.0668});
		assertEquals(rnd.N, 72);
		double[] XrayIntensities = new double[]{0.0023, 0.00112, 0.0063, 0.056, 0.035, 0.0042, 0.007, 0.00043, 0.0101, 0.0045, 0.00103, 0.0016, 3.24e-05, 0.406, 0.757, 0.0683, 0.132, 0.00121, 0.0381, 0.0058};
		assertArrayEquals(rnd.XrayIntensities, XrayIntensities);
		int[] XrayLines = new int[]{-86, -60, -89, -90, -63, -33, -34, -91, -95, -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13};
		assertArrayEquals(rnd.XrayLines, XrayLines);
		assertEquals(rnd.Z, 53);
		assertEquals(rnd.Z_xray, 52);
		assertEquals(rnd.nGammas, 1);
		assertEquals(rnd.nXrays, 20);
		assertEquals(rnd.name, "125I");
	}

	@Test
	public void test_GetRadioNuclideDataByName_125I() {
		radioNuclideData rnd = Xraylib.GetRadioNuclideDataByName("125I");
		assertEquals(rnd.A, 125);
		assertArrayEquals(rnd.GammaEnergies, new double[]{35.4919});
		assertArrayEquals(rnd.GammaIntensities, new double[]{0.0668});
		assertEquals(rnd.N, 72);
		double[] XrayIntensities = new double[]{0.0023, 0.00112, 0.0063, 0.056, 0.035, 0.0042, 0.007, 0.00043, 0.0101, 0.0045, 0.00103, 0.0016, 3.24e-05, 0.406, 0.757, 0.0683, 0.132, 0.00121, 0.0381, 0.0058};
		assertArrayEquals(rnd.XrayIntensities, XrayIntensities);
		int[] XrayLines = new int[]{-86, -60, -89, -90, -63, -33, -34, -91, -95, -68, -38, -39, -1, -2, -3, -5, -6, -8, -11, -13};
		assertArrayEquals(rnd.XrayLines, XrayLines);
		assertEquals(rnd.Z, 53);
		assertEquals(rnd.Z_xray, 52);
		assertEquals(rnd.nGammas, 1);
		assertEquals(rnd.nXrays, 20);
		assertEquals(rnd.name, "125I");
	}


	@ParameterizedTest(name="test_bad_radionuclices_names {index} -> {arguments}")
	@NullAndEmptySource
	@ValueSource(strings = {"howehfofhwf"})
	public void test_bad_radionuclices_names(String nuclide) {
		IllegalArgumentException exc = assertThrows(IllegalArgumentException.class, () -> {
			radioNuclideData rnd = Xraylib.GetRadioNuclideDataByName(nuclide);
		});
		assertEquals(exc.getMessage(), String.format("%s was not found in the radionuclide database", nuclide));
	}

	@ParameterizedTest(name="test_bad_radionuclides_indices {index} -> {arguments}")
	@ValueSource(ints = {-1, 10})
	public void test_bad_radionuclides_indices(int index) {
		IllegalArgumentException exc = assertThrows(IllegalArgumentException.class, () -> {
			radioNuclideData rnd = Xraylib.GetRadioNuclideDataByIndex(index);
		});
		assertEquals(exc.getMessage(), String.format("%d is out of the range of indices covered by the radionuclide database", index));
	}
}
